content/posts/building-images-with-img.md

```diff
@@ -0,0 +1,255 @@
+
+The current choice for many teams is Google's [kaniko](https://github.com/GoogleContainerTools/kaniko) project. This is a good option for many organizations (especially when combined with Kurt's recommendations for `PodSecurityPolicies`). However, it's worth noting that while `kaniko` itself does not require running as root, it will [require that privilege for most significant container building](https://github.com/GoogleContainerTools/kaniko#security). While this is an acceptable caveat for some organizations, it can be a deal-breaker for others.
+
+Fortunately, efforts have been underway to create an unprivileged, non-root container builder. Jessie Frazelle introduced [`img`](https://github.com/genuinetools/img), one such tool, in her blog post ["Building Container Images Securely on Kubernetes"](https://blog.jessfraz.com/post/building-container-images-securely-on-kubernetes/). I'll leave the details of design approach and usage to her post and the GitHub project page, but there are a few points I'll highlight here:
+
+1. `img` is intended to run as an arbitrary, non-root user 1000.
+2. `img` can be run without requiring the `--privileged` Docker flag or the equivalent `privileged: true` security context in Kubernetes.
+3. Syntax for building, pushing, and pulling images, among other actions, largely mirror Docker's - for example, `img build -t hello-world .` and `img push hello-world` are the commands to build and push an image called `hello-world`.
+
+In this post, we'll dive into using `img` as a `Pod` within a Google Kubernetes Engine (GKE) cluster, integrating it with Google Cloud services, and running it from a Jenkins Pipeline to automate our build and push workflow. We'll also discuss some of the security implications of `img` as they relate to running in Kubernetes.
+
+## Running `img` in Kubernetes
+Jessie's blog post includes a sample YAML manifest file (with a related [Docker container](https://r.j3ss.co/repo/img/tags)) for deploying `img` as a Kubernetes Pod, which we'll use as a starting point. However, we also want to leverage [Google Cloud's Workload Identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity) to seamlessly integrate with Google Container Registry, so we'll create a custom Docker image that includes both `img` and `gcloud`. (For a deeper dive into using Workload Identity to securely tie Kubernetes Service Accounts to cloud IAM permissions, see [part 2 of Kurt's series on best practices for Jenkins in Kubernetes](/posts/best-practices-for-cloudbees-core-jenkins-on-kubernetes/securely-using-cloud-services-from-jenkins-kubernetes-agents/).)
+
+Here's our `Dockerfile` for this combined image:
+```Dockerfile
+FROM gcr.io/cloud-builders/gcloud-slim:latest AS gcloud
+
+FROM r.j3ss.co/img:v0.5.7 AS img
+USER root
+# install Python - gcloud dependency
+RUN apk add python
+# copy google-cloud-sdk to img image
+COPY --from=gcloud /builder/google-cloud-sdk /home/user/google-cloud-sdk
+
```

Comment:
Recommend adding this:
`ENV PATH "$PATH:/home/user/google-cloud-sdk/bin/"`

Then you can use `gcloud auth configure-docker` to authenticate - no `img login` required.